pkg/controllers/clusterset/clustersetmapper/clusterpool/clusterset_clusterpool_mapper.go

```diff
@@ -0,0 +1,147 @@
+	if err := add(mgr, newReconciler(mgr, clusterSetMapper)); err != nil {
+		klog.Errorf("Failed to create ClusterSetMapper controller, %v", err)
+		return err
+	}
+	return nil
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager, clusterSetMapper *helpers.ClusterSetMapper) reconcile.Reconciler {
+	return &Reconciler{
+		client:           mgr.GetClient(),
+		scheme:           mgr.GetScheme(),
+		clusterSetMapper: clusterSetMapper,
+	}
+}
+
+// add adds a new Controller to mgr with r as the reconcile.Reconciler
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+	// Create a new controller
+	c, err := controller.New("clusterset-clusterpool-mapper-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	if err = c.Watch(&source.Kind{Type: &hivev1.ClusterPool{}},
+		&handler.EnqueueRequestForObject{}); err != nil {
+		return err
+	}
+
+	// put all clusterpool which have this clusterset label into queue if there is the clusterset event
+	err = c.Watch(&source.Kind{Type: &clusterv1alapha1.ManagedClusterSet{}},
+		&handler.EnqueueRequestsFromMapFunc{
+			ToRequests: handler.ToRequestsFunc(func(obj handler.MapObject) []reconcile.Request {
+				if _, ok := obj.Object.(*clusterv1alapha1.ManagedClusterSet); !ok {
+					// not a clusterset, returning empty
+					klog.Error("clusterset handler received non-clusterset object")
+					return []reconcile.Request{}
+				}
+
+				clusterpoolList := &hivev1.ClusterPoolList{}
+
+				//List Clusterset related cluster
+				labelSelector := metav1.LabelSelector{MatchLabels: map[string]string{
+					ClusterSetLabel: obj.Meta.GetName(),
+				}}
+				selector, err := metav1.LabelSelectorAsSelector(&labelSelector)
+				if err != nil {
+					return nil
+				}
+
+				err = mgr.GetClient().List(context.TODO(), clusterpoolList, &client.ListOptions{LabelSelector: selector})
+				if err != nil {
+					klog.Errorf("failed to list clusterpool %v", err)
+				}
+
+				var requests []reconcile.Request
+				for _, clusterpool := range clusterpoolList.Items {
+					requests = append(requests, reconcile.Request{
+						NamespacedName: types.NamespacedName{
+							Namespace: clusterpool.Namespace,
+							Name:      clusterpool.Name,
+						},
+					})
+				}
+
+				klog.V(5).Infof("List clusterpool %+v", requests)
+				return requests
+			}),
+		})
+	if err != nil {
+		return nil
+	}
+	return nil
+}
+
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+	clusterpool := &hivev1.ClusterPool{}
+	klog.V(5).Infof("reconcile: %+v", req)
+	err := r.client.Get(ctx, req.NamespacedName, clusterpool)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			r.clusterSetMapper.DeleteObjectInClusterSet(utils.ResourceNamespacedName(clusterPool, req.Namespace, req.Name))
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	if _, ok := clusterpool.Labels[ClusterSetLabel]; !ok {
```

Comment:
since claim/deployment/pool are pretty the same, i feel you could have a generic controller and  inject special logic in  it.